server/cluster/cluster.go

```diff
@@ -606,6 +606,23 @@ func (c *RaftCluster) HandleStoreHeartbeat(stats *pdpb.StoreStats) error {
 	return nil
 }
 
+// processBucketHeartbeat update the bucket information.
+func (c *RaftCluster) processBucketHeartbeat(buckets *metapb.Buckets) error {
+	region := c.core.GetRegion(buckets.GetRegionId())
+	if region == nil {
+		bucketEventCounter.WithLabelValues("region_cache_miss").Inc()
+		return errors.Errorf("region %v not found", buckets.GetRegionId())
+	}
+
+	// region should not update if the version of the buckets is less than the old one.
+	if old := region.GetBuckets(); old != nil && old.Version >= buckets.Version {
```

Comment:
Shouldn't it be `old.Version > buckets.Version`?

Reply:
region will not update if version is not change.

Reply:
It seems the region heartbeat only checks `<` here. As for a region, the version only changes when splitting or merging happens, so it's still necessary to update other statistical information while the version remains the same. I assume a bucket should also work like that, right?

https://github.com/tikv/pd/blob/f605a2cb030cb03376e0f927b1fdc0bc8b60c488/server/core/basic_cluster.go#L362

Reply:
Is there a possibility that we have an ABA problem here? Consider the following cases: 
There are two requests coming at the same time. Both request A and request B have passed the version check, A is a little bit earlier than B. But B which has a larger version is finished earlier than A. After A is finished, B is lost.

Reply:
> It seems the region heartbeat only checks `<` here. As for a region, the version only changes when splitting or merging happens, so it's still necessary to update other statistical information while the version remains the same. I assume a bucket should also work like that, right?
> 
> https://github.com/tikv/pd/blob/f605a2cb030cb03376e0f927b1fdc0bc8b60c488/server/core/basic_cluster.go#L362

the version of the bucket maybe happened at cron job or wirte skew, so it should drop the buckets info if the version is same.

Reply:
> Is there a possibility that we have an ABA problem here? Consider the following case: There are two requests coming at the same time. Both request A and request B have passed the version check, A is a little bit earlier than B. But B which has a larger version is finished earlier than A. After A is finished, B is lost.

yes, the conor will be happened , I will use CAS to avoid the latest buckets to be replaced.